ompi/mca/pml/ucx/pml_ucx_datatype.c

```diff
@@ -135,20 +161,57 @@ int mca_pml_ucx_datatype_attr_del_fn(ompi_datatype_t* datatype, int keyval,
 {
     ucp_datatype_t ucp_datatype = (ucp_datatype_t)attr_val;
 
+#ifdef HAVE_UCP_REQUEST_PARAM_T
+    free((void*)datatype->pml_data);
+#else
     PML_UCX_ASSERT((uint64_t)ucp_datatype == datatype->pml_data);
-
+#endif
     ucp_dt_destroy(ucp_datatype);
     datatype->pml_data = PML_UCX_DATATYPE_INVALID;
     return OMPI_SUCCESS;
 }
 
+#ifdef HAVE_UCP_REQUEST_PARAM_T
+__opal_attribute_always_inline__ static
+pml_ucx_datatype_t *mca_pml_ucx_init_nbx_datatype(ucp_datatype_t ucp_datatype,
+                                                  size_t size,
+                                                  int is_contig_pow2)
+{
+    pml_ucx_datatype_t *pml_datatype;
+
+    pml_datatype = malloc(sizeof(*pml_datatype));
+    if (pml_datatype == PML_UCX_DATATYPE_INVALID) {
+        PML_UCX_ERROR("Failed to allocate datatype structure");
+        ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
+    }
+
+    *pml_datatype          = pml_ucx_datatype_tmpl;
+    pml_datatype->datatype = ucp_datatype;
+    if (is_contig_pow2) {
+        pml_datatype->size_shift = (int)(log(size) / log(2.0)); /* log2(size) */
+    } else {
+        pml_datatype->size_shift = 0;
+        PML_UCX_DATATYPE_SET_VALUE(pml_datatype, op_attr_mask |= UCP_OP_ATTR_FIELD_DATATYPE);
+        PML_UCX_DATATYPE_SET_VALUE(pml_datatype, datatype = ucp_datatype);
+    }
+
+    return pml_datatype;
+}
+#endif
+
 ucp_datatype_t mca_pml_ucx_init_datatype(ompi_datatype_t *datatype)
 {
     ucp_datatype_t ucp_datatype;
     ucs_status_t status;
     ptrdiff_t lb;
     size_t size;
     int ret;
+#ifdef HAVE_UCP_REQUEST_PARAM_T
```

Comment:
can we move all things related to nbx api inside mca_pml_ucx_init_nbx_datatype? so no more ifdefs in the upper level function, mca_pml_ucx_init_datatype